### ISSUE TITLE
Show warning when turn restrictions get broken

### ISSUE THREAD

**Issue by easbar** [openstreetmap/iD#9346]
### Description

When nodes or ways that are used as members of relations with `type=restriction` are edited or deleted these relations can become broken. Maybe this is something that should be validated, so a warning will be shown if restrictions are broken?

According to the [wiki](https://wiki.openstreetmap.org/wiki/Relation:restriction) such validation should check that every relation of `type=restriction`:

* has one or more members with role 'from' and type 'way'
* has one or more members with role 'to' and type 'way'
* has exactly one member with role 'via' and type 'node' or one or more members with role 'via' and type 'way'

If we wanted to go further also these things could be validated:

* there should be no members with roles other than 'from', 'to' and 'via', which means also no members with an empty role. the only exception might be the `location_hint` role, but it is already deprecated in the wiki
* there should never be more than one member with role 'from' **and** more than one member with role 'to'
* all members with role 'via' must be connected with each other if they are ways
* probably easy: all members with role 'from' and 'to' must be connected with the member with role 'via', if it is a node. the via node must be the first or last node of the from/to member ways (ways must be split at the via node).
* probably harder: all members with role 'from' and to 'to' must be connected with the via members if they are ways.

I found about 20000 restriction relations that have no from or to member in the current planet OSM file (this seems to be the most frequent issue). Many I looked at used to have these members, but then they were deleted, because someone edited some ways in that area, and often used iD for this.

If you think this is useful I could try to provide a PR that does this.

**Comment by easbar** [comment]
I just fixed a couple more turn restrictions that got broken when users that weren't aware of these relations were editing ways:

https://www.openstreetmap.org/changeset/128422011
https://www.openstreetmap.org/changeset/128421655
https://www.openstreetmap.org/changeset/128421455

I think there are thousands of them and just displaying a warning to make people aware would probably help a lot.

**Comment by easbar** [comment]
Any comments on this?

**Comment by 1ec5** [comment]
> there should be no members with roles other than 'from', 'to' and 'via', which means also no members with an empty role

> there should never be more than one member with role 'from' and more than one member with role 'to'
all members with role 'via' must be connected with each other if they are ways

There’s an existing validation rule for empty roles in a multipolygon relation, with one-click options to add `inner` or `outer` roles:

https://github.com/openstreetmap/iD/blob/49492a0e06aef71f90e96ab3ed50098df0f4ce93/modules/validations/missing_role.js#L13-L29

There have been similar requests to restrict other relation types to a set of accepted roles, such as #8268 for multipolygons and #8709 for boundaries. ideditor/schema-builder#134 would make it easier to manage all these rules about which roles are allowed and how many times. Even better, iD would be able to show a localized name for each role to the user. Currently, non-English-speakers just have to know what `from` and `to` mean.

> probably easy: all members with role 'from' and 'to' must be connected with the member with role 'via', if it is a node. the via node must be the first or last node of the from/to member ways (ways must be split at the via node).

It isn’t quite the same thing, but iD has some safety measures when performing certain operations that might affect a restriction relation. For example, if you attempt to connect the `from` way to the `to` way at a second location, iD will refuse to connect them and they’ll scoot back to their original positions, with an error message about damaging the relation.

https://github.com/openstreetmap/iD/blob/49492a0e06aef71f90e96ab3ed50098df0f4ce93/modules/modes/drag_node.js#L227 https://github.com/openstreetmap/iD/blob/49492a0e06aef71f90e96ab3ed50098df0f4ce93/modules/modes/drag_node.js#L277-L295

This safeguard probably prevents the majority of breakage, but there’s room for improvement. (As a reviewer or data consumer, you only see the damage that makes it past this safeguard.) Note that more rules along these lines will make relation editing more rigid, so that if you know what you’re doing and need to refactor a complex relation situation, you might need to delete and recreate the relation, which unfortunately will lose history. By contrast, validation rules have the advantage that you can address them at any time before you upload, but it is easier to ignore validation warnings.

**Comment by easbar** [comment]
Thanks for you response!

> By contrast, validation rules have the advantage that you can address them at any time before you upload, but it is easier to ignore validation warnings.

Ok, I see. Wouldn't it be a good start to show a warning for some of the most common cases, namely the missing `from` and `to` members? Similar to how it is already done for multipolygon relations? And do you think it is necessary to wait for https://github.com/ideditor/schema-builder/issues/134 before doing something like this?

**Comment by 1ec5** [comment]
Yes, it would be a good start, and no, it doesn’t need to wait for schema-builder. All it means is that you’ll need to write more custom validation code directly in iD rather than tweaking a JSON file.
